pointcloud/pointcloud_file.go

```diff
@@ -538,69 +574,115 @@ func readPCDHelper(inRaw io.Reader, pctype PCType) (PointCloud, error) {
-	for i := 0; i < int(header.points); i++ {
-		line, err := in.ReadString('\n')
+func extractPCDPointASCII(in *bufio.Reader, header pcdHeader, i int) (PointAndData, error) {
+	line, err := in.ReadString('\n')
+	if err != nil {
+		return PointAndData{}, err
+	}
+	line = strings.TrimSpace(line)
+	tokens := strings.Split(line, " ")
+	if len(tokens) != int(header.fields) {
+		return PointAndData{}, fmt.Errorf("unexpected number of fields in point %d", i)
+	}
+	point := make([]float64, len(tokens))
+	for j, token := range tokens {
+		point[j], err = strconv.ParseFloat(token, 64)
 		if err != nil {
-			return nil, err
-		}
-		line = strings.TrimSpace(line)
-		tokens := strings.Split(line, " ")
-		if len(tokens) != int(header.fields) {
-			return nil, fmt.Errorf("unexpected number of fields in point %d", i)
+			return PointAndData{}, fmt.Errorf("invalid point %d field %s: %w", i, token, err)
 		}
-		point := make([]float64, len(tokens))
-		for j, token := range tokens {
-			point[j], err = strconv.ParseFloat(token, 64)
-			if err != nil {
-				return nil, fmt.Errorf("invalid point %d field %s: %w", i, token, err)
-			}
-		}
-		pcPoint, data, err := readSliceToPoint(point, header)
+	}
+	pcPoint, data, err := readSliceToPoint(point, header)
+	if err != nil {
+		return PointAndData{}, err
+	}
+
+	return PointAndData{P: pcPoint, D: data}, nil
+}
+
+func readPCDASCII(in *bufio.Reader, header pcdHeader, pc PointCloud) (PointCloud, error) {
+	for i := 0; i < int(header.points); i++ {
+		pd, err := extractPCDPointASCII(in, header, i)
 		if err != nil {
 			return nil, err
 		}
-		err = pc.Set(pcPoint, data)
+		err = pc.Set(pd.P, pd.D)
 		if err != nil {
 			return nil, err
 		}
 	}
 	return pc, nil
 }
 
-func readPCDBinary(in *bufio.Reader, header pcdHeader, pc PointCloud) (PointCloud, error) {
-	var err error
+func getPCDMetaDataASCII(in bufio.Reader, header pcdHeader) (MetaData, error) {
+	meta := NewMetaData()
 	for i := 0; i < int(header.points); i++ {
-		pointBuf := make([]float64, 3)
-		colorData := NewBasicData()
-		for j := 0; j < 3; j++ {
-			buf, err := readBuffer(in, header, j)
-			if errors.Is(err, io.EOF) {
-				break
-			}
-			pointBuf[j] = readFloat(binary.LittleEndian.Uint32(buf))
+		pd, err := extractPCDPointASCII(&in, header, i)
+		if err != nil {
+			return MetaData{}, err
+		}
+		meta.Merge(pd.P, pd.D)
+	}
+	return meta, nil
+}
+
+func extractPCDPointBinary(in *bufio.Reader, header pcdHeader) (PointAndData, error) {
+	var err error
+	pointBuf := make([]float64, 3)
+	colorData := NewBasicData()
+	for j := 0; j < 3; j++ {
+		buf, err := readBuffer(in, header, j)
+		if errors.Is(err, io.EOF) {
+			break
 		}
-		point := r3.Vector{X: 1000. * pointBuf[0], Y: 1000. * pointBuf[1], Z: 1000. * pointBuf[2]}
+		if err != nil {
+			return PointAndData{}, err
+		}
+		pointBuf[j] = readFloat(binary.LittleEndian.Uint32(buf))
+	}
+	point := r3.Vector{X: 1000. * pointBuf[0], Y: 1000. * pointBuf[1], Z: 1000. * pointBuf[2]}
```

Comment:
Can we please document somewhere that this is needed as the PCD file uses meter units & viam's Pointcloud data-structure uses millimeter units?